src/backend/distributed/utils/resource_lock.c

```diff
@@ -1196,3 +1245,169 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
 		SendCommandToWorker(nodeName, nodePort, lockCommand);
 	}
 }
+
+
+/*
+ * MakeLockRelationRecord makes a LockRelationRecord using the relation oid
+ * and the inh boolean while properly allocating the structure
+ */
+static LockRelationRecord *
+MakeLockRelationRecord(Oid relationId, bool inh)
+{
+	LockRelationRecord *lockRelationRecord = palloc(sizeof(LockRelationRecord));
+	lockRelationRecord->relationId = relationId;
+	lockRelationRecord->inh = inh;
+	return lockRelationRecord;
+}
+
+
+/*
+ * MakeLockRelationRecordList makes a list of LockRelationRecord from
+ * a list of relation oid and an inh bool which will be applied across
+ * all LockRelationRecords
+ */
+static List *
+MakeLockRelationRecordList(List *relationOidList, bool inh)
+{
+	List *result = NIL;
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationOidList)
+	{
+		LockRelationRecord *record = MakeLockRelationRecord(relationId, inh);
+		result = lappend(result, (void *) record);
+	}
+
+	return result;
+}
+
+
+/*
+ * AcquireDistributedLockOnRelations filters relations before passing them to
+ * AcquireDistributedLockOnRelations_Internal to acquire the locks.
+ *
+ * Only tables, views, and foreign tables can be locked with this function. Other relations
+ * will cause an error.
+ *
+ * Skips non distributed relations.
+ * configs parameter is used to configure what relation will be locked and if the lock
+ * should throw an error if it cannot be acquired immediately
+ */
+void
+AcquireDistributedLockOnRelations(List *relationList, LOCKMODE lockMode, uint32 configs)
+{
+	if (!ClusterHasKnownMetadataWorkers() || !EnableMetadataSync || !EnableDDLPropagation)
+	{
+		return;
+	}
+
+	/* used to store the relations that will be locked */
+	List *distributedRelationRecordsList = NIL;
+
+	/* used to ensure uniqueness of the relations that will be locked */
+	List *uniqueDistributedRelationList = NIL;
```

Comment:
Although we have a significant improvement, I still find keeping two lists hard to follow, especially given one is unique the other is not.

Can't we implement uniqueness checks on distributedRelationRecordsList? See some examples via `TaskListMember` / `TaskListDifference` / `TasksEqual` etc on list operations specific for a type